pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -204,32 +269,251 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
+
+		if _, err, updateOwnedLeaseFailed := updateLease(r.client, node, lease, time.Now(), LeaseDurationInSeconds); err != nil {
+				return err, updateOwnedLeaseFailed
+		}
 	}
 
-	return nil
+	return nil, false
+}
+
+
+func makeInt32(val int32) (*int32) {
+	tmpVal := val
+	return &tmpVal
+}
+
+func makeTimeNow(time time.Time) (*metav1.MicroTime) {
+
+	timeNow := metav1.MicroTime{Time: time}
+	return &timeNow
+}
+
+func makeExpectedOwnerOfLease(node *corev1.Node) (*metav1.OwnerReference) {
+	return &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       node.ObjectMeta.Name,
+		UID:        node.ObjectMeta.UID,
+	}
+}
+
+
+func (r *ReconcileNodeMaintenance) createOrGetExistingLease(node *corev1.Node, durationInSeconds int32) (*coordv1beta1.Lease, bool, error) {
+
+	holderIdentity := LeaseHolderIdentity
+	owner := makeExpectedOwnerOfLease(node)
+	microTimeNow :=  makeTimeNow(time.Now())
+
+	lease := &coordv1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            node.ObjectMeta.Name,
+			Namespace:       LeaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1beta1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: makeInt32(durationInSeconds),
+			AcquireTime:          microTimeNow,
+			RenewTime:            microTimeNow,
+			LeaseTransitions:     makeInt32(0),
+		},
+	}
+
+	if err := r.client.Create(context.TODO(), lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+				nodeName := node.ObjectMeta.Name
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return  nil, false, err
+				}
+				return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false,  nil
+}
+
+func needUpdateOwnedLease(lease *coordv1beta1.Lease, currentTime time.Time) (bool,bool) {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		log.Info("empty renew time or duration in sec")
+		return true, true
+	}
+	dueTime := lease.Spec.RenewTime.Time
+	duration := *lease.Spec.LeaseDurationSeconds
+	dueTime = dueTime.Add(time.Duration(time.Duration(duration) * time.Second))
+
+	deadline := currentTime
+
+	// if lease expired right now  then both update the lease and the acquire time (second rvalue)
+	// if the acquire time has been previously nil
+	if dueTime.Before(deadline) {
+		return true, lease.Spec.AcquireTime == nil
+	}
+
+	deadline = deadline.Add(time.Duration(2 * drainerTimeoutInMinutes * time.Minute))
+
+	// about to expire, update the lease but no the acquire time (second rvalue)
+	return dueTime.Before(deadline), false
+}
+
+func isValidLease(lease *coordv1beta1.Lease, currentTime time.Time) bool {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		return false
+	}
+
+	leaseDuration := time.Duration(int64(*lease.Spec.LeaseDurationSeconds) * int64(time.Second))
```

Comment:
you can skip the int64 casts
```suggestion
	leaseDuration := time.Duration(*lease.Spec.LeaseDurationSeconds) * time.Second
```

Reply:
lease.Spec.LeaseDurationSeconds is int32 * so this does not compile without a cast

Reply:
done

Reply:
strange, it works without a temp var for me like suggested :man_shrugging: 
but it's ok like it is now